hazelcast/src/test/java/com/hazelcast/internal/util/ParallelOperationInvokerTest.java

```diff
@@ -0,0 +1,279 @@
+        instance1 = factory.newHazelcastInstance(config);
+        instance2 = factory.newHazelcastInstance(config);
+        instance3 = factory.newHazelcastInstance(config);
+    }
+
+    @After
+    public void teardown() {
+        factory.terminateAll();
+    }
+
+    @Test
+    public void testInvoke() {
+        Node node = getNode(instance1);
+        Collection<UUID> uuids = invokeOnStableClusterParallel(
+                node.getNodeEngine(),
+                NoOpOperation::new,
+                0
+        ).join();
+
+        Collection<UUID> expectedUuids = getExpectedUuids(instance1, instance2, instance3);
+        assertThat(uuids)
+                .containsExactlyInAnyOrderElementsOf(expectedUuids);
+    }
+
+    @Test
+    public void testInvoke_withSingleMember() {
+        instance2.shutdown();
+        instance3.shutdown();
+
+        assertClusterSizeEventually(1, instance1);
+
+        Node node = getNode(instance1);
+        Collection<UUID> uuids = invokeOnStableClusterParallel(
+                node.getNodeEngine(),
+                NoOpOperation::new,
+                0
+        ).join();
+
+        Collection<UUID> expectedUuids = getExpectedUuids(instance1);
+        assertThat(uuids)
+                .containsExactlyInAnyOrderElementsOf(expectedUuids);
+    }
+
+    @Test
+    public void testInvoke_withThrowingTerminalException() {
+        Node node = getNode(instance2);
+        assertThatThrownBy(() -> {
+            invokeOnStableClusterParallel(
+                    node.getNodeEngine(),
+                    () -> new ExceptionThrowingOperation(new RuntimeException("expected")),
+                    0
+            ).join();
+        }).isInstanceOf(CompletionException.class)
+                .hasCauseInstanceOf(RuntimeException.class)
+                .hasStackTraceContaining("expected");
+    }
+
+    @Test
+    public void testInvoke_withThrowingCompletionExceptionWithNullCause() {
+        Node node = getNode(instance2);
+        assertThatThrownBy(() -> {
+            invokeOnStableClusterParallel(
+                    node.getNodeEngine(),
+                    () -> new ExceptionThrowingOperation(new CompletionException(null)),
+                    0
+            ).join();
+        }).isInstanceOf(CompletionException.class)
+                .hasCause(null);
+    }
+
+    @Test
+    public void testInvoke_withThrowingIgnorableException() {
+        Node node = getNode(instance3);
+        Collection<UUID> uuids = invokeOnStableClusterParallel(
+                node.getNodeEngine(),
+                () -> new ExceptionThrowingOperation(new HazelcastInstanceNotActiveException("expected")),
+                2
+        ).join();
+
+        Collection<UUID> expectedUuids = getExpectedUuids(instance1, instance2, instance3);
+        assertThat(uuids)
+                .containsExactlyInAnyOrderElementsOf(expectedUuids);
+    }
+
+    @Test
+    public void testInvoke_withThrowingTopologyChangeException() {
```

Comment:
we can maybe add a test about maxRetryCount. If it is 3, max retry should be 3. I think we can achieve that by a spy.

Reply:
I can, but then I would have to make some internal parts of the invoker package private just for tests, and I don't really want that. I have changed the `testInvoke_withThrowingTopologyChangeException` a bit, I hope this would be enough. If you want to see such a test, I could add it though, at the expense of making some things not private